ros2param/ros2param/verb/dump.py

```diff
@@ -0,0 +1,145 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+import yaml
+
+from rcl_interfaces.msg import ParameterType
+from rcl_interfaces.srv import ListParameters
+import rclpy
+from ros2cli.node.direct import DirectNode
+from ros2cli.node.strategy import add_arguments
+from ros2cli.node.strategy import NodeStrategy
+from ros2node.api import get_absolute_node_name
+from ros2node.api import get_node_names
+from ros2node.api import NodeNameCompleter
+from ros2node.api import parse_node_name
+from ros2param.api import call_get_parameters
+from ros2param.verb import VerbExtension
+from ros2service.api import get_service_names
+
+
+class DumpVerb(VerbExtension):
+    """Dump the parameters to a yaml file."""
+
+    def add_arguments(self, parser, cli_name):  # noqa: D102
+        add_arguments(parser)
+        arg = parser.add_argument(
+            'node_name', help='Name of the ROS node')
+        arg.completer = NodeNameCompleter(
+            include_hidden_nodes_key='include_hidden_nodes')
+        parser.add_argument(
+            '--include-hidden-nodes', action='store_true',
+            help='Consider hidden nodes as well')
+        parser.add_argument(
+            '--file-path',
+            default='.',
+            help='The absolute path were to save the generated file')
+
+    @staticmethod
+    def get_parameter_value(node, node_name, param):
+        response = call_get_parameters(
+            node=node, node_name=node_name,
+            parameter_names=[param])
+
+        # requested parameter not set
+        if not response.values:
+            return '# Parameter not set'
+
+        # extract type specific value
+        pvalue = response.values[0]
+        if pvalue.type == ParameterType.PARAMETER_BOOL:
+            value = pvalue.bool_value
+        elif pvalue.type == ParameterType.PARAMETER_INTEGER:
+            value = pvalue.integer_value
+        elif pvalue.type == ParameterType.PARAMETER_DOUBLE:
+            value = pvalue.double_value
+        elif pvalue.type == ParameterType.PARAMETER_STRING:
+            value = pvalue.string_value
+        elif pvalue.type == ParameterType.PARAMETER_BYTE_ARRAY:
+            value = pvalue.byte_array_value
+        elif pvalue.type == ParameterType.PARAMETER_BOOL_ARRAY:
+            value = pvalue.bool_array_value
+        elif pvalue.type == ParameterType.PARAMETER_INTEGER_ARRAY:
+            value = pvalue.integer_array_value
+        elif pvalue.type == ParameterType.PARAMETER_DOUBLE_ARRAY:
+            value = pvalue.double_array_value
+        elif pvalue.type == ParameterType.PARAMETER_STRING_ARRAY:
+            value = pvalue.string_array_value
+        elif pvalue.type == ParameterType.PARAMETER_NOT_SET:
+            value = None
+        else:
+            value = None
+
+        return value
+
+    def insert_dict(self, dict, key, value):
+        split = key.split("/", 1)
```

Comment:
I'm not sure if namespaces are delimited with `/` or `.` as I have seen both :s .

Reply:
Depends what you mean by namespace.

A node namespace (the thing set with `__ns:=/foo/bar`) has tokens delimited with `/`.  Multiple levels in a parameter name are delimited with a `.`. You can get the character with this import [`from rclpy.parameter import PARAMETER_SEPARATOR_STRING`](https://github.com/ros2/rclpy/blob/a28b8258c2a1f6300bb407b2d7e10e4e98b60b48/rclpy/rclpy/parameter.py#L20)

```
$ ros2 run demo_nodes_cpp parameter_blackboard  __ns:=/foo/bar &
[1] 25174
...
$ ros2 node list
/foo/bar/parameter_blackboard
$ ros2 param list /foo/bar/parameter_blackboard
  use_sim_time
$ ros2 param set /foo/bar/parameter_blackboard ping.pong 5
Set parameter successful
$ ros2 param list /foo/bar/parameter_blackboard
  ping.pong
  use_sim_time
```

If you're dumping to a yaml file, [you can just use indentation](https://github.com/ros2/rcl/blob/b6f4bc97faf34dd85f7acff7cb27945cfe438369/rcl_yaml_param_parser/test/multi_ns_correct.yaml#L4-L23).


```
$ ros2 run demo_nodes_cpp parameter_blackboard  __ns:=/lidar_ns/lidar_1/lidar11 __node:=lidar111 __params:=/home/sloretz/ws/ros2/src/ros2/rcl/rcl_yaml_param_parser/test/multi_ns_correct.yaml &
[1] 25358
...
$ ros2 param list /lidar_ns/lidar_1/lidar11/lidar111
  driver1.driver11.bk_sensor_specs
  driver1.driver11.driver111.dx
  driver1.driver11.driver111.dy
  driver1.driver11.driver111.fr_sensor_specs
  driver2.driver21.dx
  driver2.driver21.dy
  id
  is_front
  name
  ports
  use_sim_time
```

Reply:
I did mean "Multiple levels in a parameter name", didn't know about `PARAMETER_SEPARATOR_STRING` thanks.  
This new verb does dump into a yaml file using multi-level indentation for multi-level parameters, see the example in my very first comment.

Reply:
But how does one declare those multi-level parameters in python ?? The following does not work,
```python
node.declare_parameter('foo.str_param', 'foo')
```
for `topic name must not contain characters other than alphanumerics, '_', '~', '{', or '}'`

Reply:
That looks like a bug

https://github.com/ros2/rclpy/blob/a28b8258c2a1f6300bb407b2d7e10e4e98b60b48/rclpy/rclpy/validate_parameter_name.py#L31-L33

@jubeira FYI

Reply:
@artivis Would you be willing to open an issue about `declare_parameter('foo.str_param', 'foo')` not working on the rclpy repo? https://github.com/ros2/rclpy/issues/new

Reply:
Looks like an issue has been opened https://github.com/ros2/rclpy/issues/373

Reply:
Ok great, thanks.